GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/IPC/VFSMessageListener.m

```diff
@@ -0,0 +1,133 @@
+#import "VFSMessageParser.h"
+
+NSString * const NotificationServerPipeName = @"vfsforgit.notification";
+
+@interface VFSMessageListener ()
+
+@property (strong) NSFileHandle *connectionHandle;
+@property CFSocketRef socketRef;
+
+@end
+
+@implementation VFSMessageListener
+
+- (instancetype)initWithSocket:(NSString *)socketPath
+                      callback:(nonnull NewMessageCallback)callback
+{
+    if (self = [super init])
+    {
+        _socketPath =
+        [[socketPath stringByAppendingPathComponent:NotificationServerPipeName]
+         copy];
+        _messageCallback = [callback copy];
+    }
+    
+    return self;
+}
+
+- (BOOL)startListening
+{
+    CFSocketRef cfSocket;
+    if ((cfSocket = CFSocketCreate(kCFAllocatorDefault,
+                                   PF_LOCAL,
+                                   SOCK_STREAM,
+                                   0,
+                                   kCFSocketNoCallBack,
+                                   NULL,
+                                   NULL)) == NULL)
+    {
+        return NO;
+    }
+    
+    CFAutorelease(cfSocket);
+    
+    int reuse = TRUE;
+    int socketDescriptor = CFSocketGetNative(cfSocket);
+    if (setsockopt(socketDescriptor,
+                   SOL_SOCKET,
+                   SO_REUSEADDR,
+                   (void *) &reuse,
+                   sizeof(reuse)))
+    {
+        return NO;
+    }
+    
+    if ([[NSFileManager defaultManager] fileExistsAtPath:self.socketPath] &&
+        ![[NSFileManager defaultManager] removeItemAtPath:self.socketPath error:nil])
+    {
+        return NO;
+    }
+    
+    struct sockaddr_un sockAddress = {};
+    memset(&sockAddress, 0, sizeof(sockAddress));
+    sockAddress.sun_family = AF_UNIX;
+    sockAddress.sun_len = sizeof(sockAddress);
+    [self.socketPath getCString:sockAddress.sun_path
+                      maxLength:sizeof(sockAddress.sun_path)
+                       encoding:NSUTF8StringEncoding];
+    
+    CFDataRef addressData = CFDataCreate(kCFAllocatorDefault,
+                                         (const UInt8 *) &sockAddress,
+                                         sizeof(sockAddress));
+    CFAutorelease(addressData);
+    if (CFSocketSetAddress(cfSocket, addressData) != kCFSocketSuccess)
+    {
+        return NO;
+    }
+    
+    self.socketRef = (CFSocketRef) CFRetain(cfSocket);
+    
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(connectionCallback:)
+                                                 name:NSFileHandleConnectionAcceptedNotification
+                                               object:nil];
+    
+    self.connectionHandle = [[NSFileHandle alloc] initWithFileDescriptor:socketDescriptor
+                                                          closeOnDealloc:YES];
+    [self.connectionHandle acceptConnectionInBackgroundAndNotify];
+    
+    return YES;
+}
+
+- (void)stopListening
```

Comment:
Add a -dealloc method to ensure the cleanup happens correctly on termination or whenever the object is released. Otherwise you are relying on the caller to call this appropriately. It may be fine right now because stopListening is called on application terminate, but could change with code refactoring etc.

Reply:
yep... added.

Couple notes
1. `[super dealloc]` is not added because of `ARC`.
2. [Apple docs](https://developer.apple.com/documentation/foundation/nsnotificationcenter/1413994-removeobserver?language=objc) suggest it is not required to un-register notification observers from inside a `dealloc` method.